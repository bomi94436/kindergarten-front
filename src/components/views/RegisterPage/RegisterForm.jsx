import React, { useEffect, useState } from "react";
import styled from "styled-components";
import { makeStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";

import SearchKindergarten from "../../../containers/RegisterPage/SearchKindergartenContainer";
import { contents } from "./registerConents";
import { isEveryFieldValid } from "../../../utils/validation";
import { FormInput, FormButton, FormDropdown } from "./RegisterFormComp";
import AddStudent from "../../../containers/RegisterPage/AddStudentContainer";
import "../../../utils/styles.css";

const FormCover = styled.div`
  max-width: 50%;
  margin: 1rem auto 0;
  padding: 20px;
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
`;

const StyledButton = styled(Button)`
  height: 4rem;
  flex-grow: 1;
  font-size: 1.2rem;
  border-radius: 30px;
`;

const useStyles = makeStyles((theme) => ({
  root: {
    "& > *": {
      margin: theme.spacing(0.5),
      display: "flex",
    },
  },
}));

const RegisterForm = ({
  history,
  state,
  setRegister,
  setRegisterValid,
  postRegister,
  getExistId,
}) => {
  const [title, setTitle] = useState("");
  const classes = useStyles();

  useEffect(() => {
    const path = history.location.pathname.split("/")[2];
    if (path === "user") {
      setTitle("학부모(일반회원)");
      setRegisterValid({ actor: "user", act: "set" });
    } else if (path === "teacher") {
      setTitle("선생님");
      setRegisterValid({ actor: "teacher", act: "set" });
    } else if (path === "director") {
      setTitle("원장님");
      setRegisterValid({ actor: "director", act: "set" });
    }
  }, [history.location.pathname, setRegisterValid]);

  const handleSubmit = (e) => {
    const path = history.location.pathname.split("/")[2];
    e.preventDefault();

    const dataToSubmit = {
      userid: state.value.userid,
      password: state.value.password,
      name: state.value.name,
      phone: state.value.phone,
      email: `${state.value.firstEmail}@${state.value.lastEmail}`,
      role: path.toUpperCase(),
    };

    if (path === "teacher" || path === "director") {
      dataToSubmit.kindergarten_id = state.search.selected.id;
    } else if (path === "user" && state.students) {
      dataToSubmit.student = [];
      state.students.forEach((student) => {
        let studentData = {
          kindergarten_id: student.value.kindergarten_id,
          name: student.value.name,
          year: student.value.date.getFullYear().toString(),
          month: (student.value.date.getMonth() + 1).toString(),
          day: student.value.date.getDate().toString(),
        };
        dataToSubmit.student.push(studentData);
      });
    }

    postRegister(dataToSubmit).then((res) => {
      alert(res.msg);
      if (res.success) {
        history.push("/");
      }
    });
  };

  return (
    <div
      style={{
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
      }}
      className="container"
    >
      <FormCover>
        <p style={{ fontSize: "2rem", fontWeight: 500 }}>{title} 회원가입</p>

        <form onSubmit={(e) => handleSubmit(e)} className={classes.root}>
          <div style={{ marginBottom: "2rem" }}>
            <FormInput
              element={contents.userid}
              setRegister={setRegister}
              state={state}
            />

            <div
              style={{
                display: "flex",
                marginLeft: "1rem",
              }}
            >
              <FormButton
                element={contents.checkDuplication}
                getExistId={getExistId}
                id={state.value.userid}
                isUseridValid={state.valid.userid}
              />
            </div>
          </div>

          <FormInput
            element={contents.password}
            setRegister={setRegister}
            state={state}
          />
          <br />

          <FormInput
            element={contents.rePassword}
            setRegister={setRegister}
            state={state}
          />
          <br />

          <FormInput
            element={contents.name}
            setRegister={setRegister}
            state={state}
          />
          <br />

          <FormInput
            element={contents.phone}
            setRegister={setRegister}
            state={state}
          />
          <br />

          <div
            style={{
              display: "flex",
              alignItems: "center",
              justifyContent: "space-between",
            }}
          >
            <FormInput
              element={contents.firstEmail}
              setRegister={setRegister}
              state={state}
            />
            <span style={{ margin: "0.5rem" }}>@</span>
            <FormInput
              element={contents.lastEmail}
              setRegister={setRegister}
              state={state}
            />
            <FormDropdown
              element={contents.emailList}
              setRegister={setRegister}
              state={state}
            />
          </div>

          {title === "선생님" || title === "원장님" ? (
            <SearchKindergarten />
          ) : (
            <AddStudent />
          )}

          <div style={{ marginTop: "3rem" }}>
            {isEveryFieldValid(state.valid, state.students) ? (
              <StyledButton type="submit" variant="contained" color="primary">
                회원가입
              </StyledButton>
            ) : (
              <StyledButton
                type="submit"
                variant="contained"
                color="primary"
                disabled
              >
                회원가입
              </StyledButton>
            )}
          </div>
        </form>
      </FormCover>
    </div>
  );
};

export default RegisterForm;
