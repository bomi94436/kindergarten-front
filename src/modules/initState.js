const initState = {
  loading: {
    POST_REGISTER: false,
    GET_EXISTID: false,
    GET_REGISTER_SEARCH: false,
  },
  register: {
    value: {
      userid: "",
      password: "",
      rePassword: "",
      name: "",
      phone: "",
      firstEmail: "",
      lastEmail: "",
    },
    valid: {
      userid: null,
      checkDuplication: {
        checked: null,
        id: null,
      },
      password: null,
      rePassword: null,
      name: null,
      phone: null,
      firstEmail: null,
      lastEmail: null,
    },
    students: [],
    search: {
      opened: false,
      type: "name",
      value: "",
      selected: {
        id: null,
        name: null,
      },
      contents: null,
      page: {
        total: null,
        current: null,
      },
    },
    student: {
      value: {
        name: "",
        date: new Date(),
        kindergarten_id: null,
      },
      valid: {
        name: null,
        date: null,
        kindergarten_id: null,
      },
      kindergarten_selected: {
        id: null,
        name: null,
      },
      // search: {
      //   opened: false,
      //   type: "name",
      //   value: "",
      //   selected: {
      //     id: null,
      //     name: null,
      //   },
      //   contents: null,
      //   page: {
      //     total: null,
      //     current: null,
      //   },
      // },
    },
  },
};

export default initState;
