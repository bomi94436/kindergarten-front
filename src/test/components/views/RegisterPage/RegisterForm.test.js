import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { createMemoryHistory } from "history";
import configureMockStore from "../../../../modules/store";
import RegisterForm from "../../../../components/views/RegisterPage/RegisterForm";
import initState from "../../../../modules/initState";
import { Provider } from "react-redux";

describe("<RegisterForm /> 에서", () => {
  const history = createMemoryHistory();
  const paths = ["user", "teacher", "director"];
  const title = [
    "학부모(일반회원) 회원가입",
    "선생님 회원가입",
    "원장님 회원가입",
  ];
  const mockChange = jest.fn();
  const mockApi = jest.fn(() => Promise.resolve("response"));

  const setup = (path) => {
    const { store } = configureMockStore();
    let state = initState.register;
    history.push(`/register/${path}`);

    return render(
      <Provider store={store}>
        <RegisterForm
          history={history}
          state={state}
          setRegister={mockChange}
          setRegisterValid={mockApi}
          postRegister={mockApi}
          getExistId={mockApi}
        />
      </Provider>
    );
  };

  it("3개의 회원가입 페이지들이 잘 렌더링된다", () => {
    let component = null;
    paths.forEach((path) => {
      component = setup(path);
    });
  });

  it("스냅샷이 일치한다", () => {
    const component = setup("user");
    expect(component.container).toMatchSnapshot();
  });

  paths.forEach((path, index) => {
    describe(`/${path} 에서`, () => {
      it(`타이틀 및 폼이 정상적으로 보여진다`, () => {
        const component = setup(path);
        component.getByText(title[index]);
        component.getByLabelText("아이디");
        component.getByText("중복 확인");
        component.getByLabelText("비밀번호");
        component.getByLabelText("비밀번호 확인");
        component.getByLabelText("이름");
        component.getByLabelText("휴대폰 번호");
        component.getByLabelText("이메일");
      });
    });
  });
});
