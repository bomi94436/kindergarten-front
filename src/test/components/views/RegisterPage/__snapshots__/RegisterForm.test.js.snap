// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterForm /> 에서 스냅샷이 일치한다 1`] = `
<div>
  <div
    style="display: flex; justify-content: center;"
  >
    <div
      class="sc-eCApGN jXAhvR"
    >
      <p
        style="font-size: 2rem; font-weight: 500;"
      >
        학부모(일반회원)
         회원가입
      </p>
      <form
        class="makeStyles-root-11"
      >
        <div
          style="margin-bottom: 2rem;"
        >
          <div
            class="MuiFormControl-root MuiTextField-root sc-hKFyIo dZgHxg"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
              for="userid outlined-basic"
              id="userid outlined-basic-label"
            >
              아이디
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="userid outlined-basic"
                placeholder="숫자, 영어 조합하여 6자리 이상 입력"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-14"
                >
                  <span>
                    아이디
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            style="display: flex; margin-left: 1rem;"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              id="checkDuplication"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                중복 확인
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root sc-hKFyIo dZgHxg"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
            for="password outlined-basic"
            id="password outlined-basic-label"
          >
            비밀번호
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              id="password outlined-basic"
              placeholder="숫자, 영어 조합하여 8자리 이상 입력"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-14"
              >
                <span>
                  비밀번호
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <br />
        <div
          class="MuiFormControl-root MuiTextField-root sc-hKFyIo dZgHxg"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
            for="rePassword outlined-basic"
            id="rePassword outlined-basic-label"
          >
            비밀번호 확인
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              id="rePassword outlined-basic"
              placeholder=""
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-14"
              >
                <span>
                  비밀번호 확인
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <br />
        <div
          class="MuiFormControl-root MuiTextField-root sc-hKFyIo dZgHxg"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
            for="name outlined-basic"
            id="name outlined-basic-label"
          >
            이름
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              id="name outlined-basic"
              placeholder="2글자 이상 입력"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-14"
              >
                <span>
                  이름
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <br />
        <div
          class="MuiFormControl-root MuiTextField-root sc-hKFyIo dZgHxg"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
            for="phone outlined-basic"
            id="phone outlined-basic-label"
          >
            휴대폰 번호
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              id="phone outlined-basic"
              placeholder="숫자만 입력"
              type="number"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-14"
              >
                <span>
                  휴대폰 번호
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <br />
        <div
          style="display: flex; align-items: center; justify-content: space-between;"
        >
          <div
            class="MuiFormControl-root MuiTextField-root sc-hKFyIo dZgHxg"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
              for="firstEmail outlined-basic"
              id="firstEmail outlined-basic-label"
            >
              이메일
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="firstEmail outlined-basic"
                placeholder=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-14"
                >
                  <span>
                    이메일
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <span
            style="margin: 0.5rem;"
          >
            @
          </span>
          <div
            class="MuiFormControl-root MuiTextField-root sc-hKFyIo dZgHxg"
          >
            
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="lastEmail outlined-basic"
                placeholder="example.com"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-13"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root makeStyles-formControl-16"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
              data-shrink="true"
              id="dropdown"
            >
              주소
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby="dropdown emailList"
                class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
                id="emailList"
                role="button"
                tabindex="0"
              >
                직접 입력
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput"
                tabindex="-1"
                value="직접 입력"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-14 PrivateNotchedOutline-legendNotched-15"
                >
                  <span>
                    주소
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="makeStyles-row-18"
          style="margin-top: 1rem;"
        >
          <span
            style="margin: 1rem; font-size: 1.2rem;"
          >
            학생 추가
          </span>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
            tabindex="0"
            type="button"
            variant="contained"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="makeStyles-root-17"
        >
          <div
            class="MuiGrid-root makeStyles-row-18 MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
            >
              <span
                class="makeStyles-label-19"
              >
                이름
              </span>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <span
                class="makeStyles-label-19"
              >
                생년월일
              </span>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <span
                class="makeStyles-label-19"
              >
                소속 유치원
              </span>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
            >
              <span
                class="makeStyles-label-19"
              >
                삭제
              </span>
            </div>
          </div>
          <p
            style="border-bottom: 1px solid gray; margin: 0.5rem;"
          />
        </div>
        <div
          style="margin-top: 3rem;"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-jSFkmK dYUkBY MuiButton-containedPrimary Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              회원가입
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
